runtime/io/common/src/aws/smithy/kotlin/runtime/io/HashingByteReadChannel.kt

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.io
+
+import aws.smithy.kotlin.runtime.hashing.HashFunction
+import aws.smithy.kotlin.runtime.util.InternalApi
+import aws.smithy.kotlin.runtime.util.encodeBase64String
+import kotlinx.coroutines.CompletableDeferred
+
+/**
+ * A channel which hashes data as it is being read
+ * @param hash The [HashFunction] to hash data with
+ * @param chan the [SdkByteReadChannel] to hash
+ * @param deferred the optional [CompletableDeferred] to be completed using the Base64-encoded digest of [chan] after it is exhausted
+ */
+@InternalApi
+public class HashingByteReadChannel(
+    private val hash: HashFunction,
+    private val chan: SdkByteReadChannel,
+    private val deferred: CompletableDeferred<String>? = null,
```

Comment:
Speaking my last peace on this and then I'll shut up.... I still don't think this is a concern of `HashingByteReadChannel` or `HashingSource`. I think this functionality should be isolated to where it's needed and be done as an additional (simple) wrapper.

Reply:
I had a bit of trouble isolating it only to where it's needed because both the request and response interceptors need to use it. I could place it in one and use it in the other, but that adds a weird linkage between them